### DISCUSSION TITLE
How to process non-php files in general

### DISCUSSION THREAD

**Discussion by AlexeyOnsite** [rectorphp/rector#6265] [Q&A] [answered]
After converting legacy php scripts to controllers, I need to replace all links to them. Some of them are in JS files. Is there a way to do it?
Currently the list of non php files extensions is hard coded, if I understand correctly?

**Comment by TomasVotruba** [comment, marked as answer]
Hi, this might help: https://github.com/rectorphp/rector/pull/6060

@sabbelasichon Hi Sebastian, could you elaborate on this topic, as an author of the PR?

**Reply by AlexeyOnsite** [reply]
I've remebered I've seen something similar, thanks for the link.
Is it already in rector-prefixed?

**Reply by sabbelasichon** [reply]
Yes, it is already part of rector-prefixed. 

You can create your custom JavaScriptFileProcessor implementing the FileProcessorInterface for your JS files. Could you give me an example of what you are trying to achieve?

**Reply by AlexeyOnsite** [reply]
That's great!
First I have to update links to php scripts like script.php -> /path/to/script.php
Second wenn all scripts are converted to controllers links have to be updated to routes.
Unfortunately not only in php files but also in js files, too.

**Reply by sabbelasichon** [reply]
Here is one example to process specific xml Files: https://github.com/sabbelasichon/typo3-rector/blob/master/src/FlexForms/FlexFormsProcessor.php#L18

So i would guess in your case it can be something like this

```php
final class JavaScriptFileProcessor implements FileProcessorInterface
{
    public function supports(File $file): bool
    {
        $smartFileInfo = $file->getSmartFileInfo();

        if (! in_array($smartFileInfo->getExtension(), $this->getSupportedFileExtensions(), true)) {
            return false;
        }

        return true;
    }

    public function process(array $files): void
    {
        foreach ($files as $file) {
            // Do something here
        }
    }

    public function getSupportedFileExtensions(): array
    {
        return ['js', 'ts'];
    }
}
```

You can then register your FileProcessor in your rector.php configuration file and it will be automatically processed. You can also inject some other classes you need in order to work with your Processor. 
If these classes are already known to the Dependency Injection Container everything is fine. If not, you have to register your helper classes in the rector configuration additionally.

**Reply by AlexeyOnsite** [reply]
Thank you very much!
I need to take a look into your typo3 stuff :)

**Comment by AlexeyOnsite** [comment]
@TomasVotruba by the way the next question, since it correlates a bit: I don't quite understand what a post-rector is. Is it something that will be invoked automatically or something I should explicit configure?

**Reply by TomasVotruba** [reply]
Good question! The post Rectors run after all the PHP ones. They intend to cleanup after all the changes, e.g. import FQN names to short ones.
